drag-and-drop.md

```diff
@@ -0,0 +1,126 @@
+
+## Background
+
+There are few issues opened and related to missing drag and drop functionality:
+* Add drag and drop support https://github.com/fyne-io/fyne/issues/142
+* Drag and drop to application icon on macOS https://github.com/fyne-io/fyne/issues/497
+
+Existing interface provide a minimum capabilities to programmatically start a drug operation and get notified when it complets.
+This functionality doesn't provide an interface to communicate between widgets within the application nor send/receive drag and drop with other applications.
+
+## Architecture / API
+
+Existing interface will not change to preserve compatibility with existing code.
+The proposed changes will extend the interface and give user missing capabilities.
+
+The **fyne.CanvasObject** wishing to initiate drag operation must implement **fyne.Draggable** interface. No interface changes.
+
+The **fyne.CanvasObject** wishing to receiver the drop must implement **fyne.DragReceiver** interface.
+
+```
+package fyne
+
+type DragReceiver interface {
+	DragDrop(Draggable)
+}
+```
+
+To pass information to receiver the **fyne.Draggable** can also implement **drag.Info** interface.
+
+```
+package drag
+
+type Info interface {
+	Payload() Payload
+}
+```
+
+To initiate a drag-and-drop outside the Fyne application the object implement **drag.AppDraggable** interface. This interface will allow the object prepare the **drag.Payload**.
+
+```
+package drag
+
+type AppDraggable interface {
+	// StartDrag return true when object ready to start drag-and-drop
+	StartDrag() bool
+	// Payload return Payload to attach to drag-and-drop session
+	Payload() Payload
+}
+```
+
+The initial implementations for **drag.Payload** will include **drag.PayloadString**, **drag.PayloadSliceOfString**. Other types can be added as needed and assuming the platform supports such type.
+
+```
+package drag
+
+type Payload interface {}
+
+type PayloadString string
+type PayloadSliceOfString []string
+
+```
+
+To receive the drag-and-drop from outside the Fyne application the **fyne.CanvasObject** must implement the **fyne.DragReceiver** interface.
+
+The **fyne.Draggable** can implement **drag.Paintable** interface to provide an image that can be used to paint a cursor during drag-and-drop operation.
+
+```
+package drag
+
+type Paintable interface {
+	CursorImage() image.Image
+}
+
+```
+
+## Implementation
+
+### Drag-and-Drop within the Fyne application
+
+The drag and drop within the Fyne application is implemented by finding the component under the mouse pointer and supporting **fyne.DragReceiver** interface in each of the windows of application sorted by z-index. Only the first window with object of **fyne.DragReceiver** will get the call with fyne.Draggable.
+
+The macOS implementation was tested to confirm this design.
+
+This require for each platform to provide a window position, based on its order from front to back among all visible application windows.
```

Comment:
I don't think this is the case - we should be able to get events into the target window and not need to know positions or orders.

Reply:
at least on macOS, the events only goes to active window of application

Reply:
I think that may need looked into rather than designed for. The OS will send drag events to any window, so we may need to explore why (if) they are not received.

Reply:
the go-gl initialization work in such way that mouse move events always sent to active windows, the inactive window get mouse move events only when cursor in a visible area and no buttons pressed. Once any button is pressed, the mouse move events sent only to the active window where it was pressed. 
The macOS seem allows drag event into inactive window (at least Finder's drop & drop show hover in other windows).
But, to allow drag movement sent to visible window will require changes (yet to be found) in go-gl.
I can't comment on how other platforms handle this.

Reply:
> the go-gl initialization work in such way that mouse move events always sent to active windows

You might be looking in the wrong place - events are handled by GLFW not go-gl.
That said I agree that something strange is happening with drags over non-source windows, perhaps we need to try and convert the drag to an OS drag if the mouse leaves a window?

The reason I think this is that drags over non-Fyne windows stop working as well, so we can't work around it with changes internal to Fyne, therefore we have to look at how the drag works rather than tracking secondary window positions and zorder.

Reply:
that is correct. only-Fye is limited to single window visually and at the moment physically. The multi-window only-Fyne is possible, but cursor will be lost and no feed back on another window (see attached).
Tap directly into a running platform APIs would require for flawless multi-window and multi-app implementation. That may require changes in GLFW, as events handled there.
However, the Fyne interface could stay the same and work can be done in two phases:

1. interface and only-Fyne implementation 
2. platform specific implementation on each supported platform

https://user-images.githubusercontent.com/298260/127319778-e328dabf-9be8-4658-b33e-551bfa6f24b4.mov

Reply:
I don't think we can split this in 2. If we find in step 2 that we need a new interface to differentiate then we'll break usage of the former API. Not a problem if we know 1 and 2 will happen in tight succession, but if we're not sure of that best to address both together.
Another reason I think this is that the complexity is in the OS specifics, so best to address that first.

Reply:
We need volunteers for iOS, Android, Windows and X11 (anything else?)
I can do macOS implementation.
Hope GLFW won't need changes. But in event the GLWF need changes, how this is handled?

Reply:
Sorry for the slow reply, I was busy with 2.0.4 and some business things.
If you can get macOS working then I can do iOS/Android/X11. Hopefully someone will contribute Windows support, but if not I can look at that too.